File: internal/analysis/funnel.go
package analysis

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// The funnel: of the people who arrived, how many got to each step.
//
// # THE ONE THING THAT MAKES IT HARD
//
// The top of the funnel is browsers and the bottom is accounts. Somebody
// arrives with no account and subscribes with one, so counting the two halves
// by the identity each event happens to carry would compare different
// populations and call the result a conversion rate.
//
// So a PERSON is defined once, here: an account if the identity is linked to
// one, and the visitor otherwise. That is what `account_visitors` is for, and
// it is the reason the visitor has an identity before the account exists at all
// (K-10).
//
// # A STEP WITH NO EVENT IS NOT A STEP WITH NOBODY
//
// ONE of the eight cannot be emitted today: subscribing, because nothing
// creates a subscription until there is a gateway. Reported as zero it would
// read as everybody dropping out — the same mistake as a discrimination index
// of zero that was never measured. It comes back saying no event exists yet,
// and a screen has to show that differently from a drop.
//
// Verifying an address was the second until `account.confirmed` started being
// emitted. It is measured now, and `cmd/seed` emits it too — a step named here
// that the seeder does not produce would put a cliff in every seeded funnel,
// which is this same mistake wearing different clothes.

// Step is one step of the funnel, with what it is counted from.
type Step struct {
	// Event is the name in the stream, or empty when nothing emits this step.
	Event string

	/* Platform says this step is not something that happens inside a school.

	   THERE IS EXACTLY ONE, AND IT IS NOT AN OVERSIGHT. One subscription covers
	   every school (N-02), so a subscription belongs to no school and its event
	   carries no tenant — `event.ForPlatform` says exactly that. Seven steps of
	   this funnel are things somebody did at a school and the eighth is not,
	   and flattening the difference would mean picking a school to attribute a
	   purchase to that was never about one.

	   WHAT THE STEP THEN MEANS is: of the people who arrived HERE, how many
	   went on to subscribe — anywhere. That is the honest reading and it is the
	   question somebody opens a funnel to ask. It also means somebody who
	   arrived at two schools and subscribed once is counted under both, which
	   is true rather than double-counted: each funnel is asking about its own
	   arrivals. It is the shape the countries screen already carries a sentence
	   about, and this screen now carries its own. */
	Platform bool

	// Label is what it is called on a screen, in English. The source language
	// is English and the key is the string (N-06).
	Label string

	// People is how many distinct people reached it.
	People int

	// Measured is false where nothing emits this step. It is a separate field
	// rather than `People == 0`, because "nobody got here" and "nothing counts
	// this" are different facts and a screen that showed them alike would
	// report a missing feature as a total drop-off.
	Measured bool

	// Why says what is missing, for a step that is not measured.
	Why string
}

// Counting says which population the funnel counts.
//
// IT IS THIS PACKAGE'S OWN WORD AND NOT THE STREAM'S. `internal/event` has a
// type of the same name with the same three values, and this one exists rather
// than importing it because a module may not import another module (X-02) —
// `cmd/api` is what says the two are the same, in one line, where the rest of
// the wiring already is.
//
// The duplication is small and it is the price of the boundary. What makes it
// safe is that BOTH sides fall back to `real` for anything they do not
// recognise: a value lost in translation narrows the population and never
// widens it into a report about people that quietly counts invented ones.
type Counting string

const (
	// CountingReal is people who came here on their own. The default, and what
	// a screen shows unless somebody asks otherwise.
	CountingReal Counting = "real"

	// CountingSeeded is the seeded population alone — what `cmd/seed` wrote.
	CountingSeeded Counting = "seeded"

	// CountingEverybody is both, for a screen that is showing a demonstration
	// and says on its face that it is.
	CountingEverybody Counting = "everybody"
)

// Reading answers whether a word is one of the three.
//
// THE FALLBACK IS SAFE AND THE REFUSAL IS HONEST, and they are different jobs.
// `Counting("everbody")` counts real people, which is the right thing for the
// SQL to do and the wrong thing for a screen to do quietly — the switch would
// say "including the seeded population" over a chart that excluded it. So a
// caller that took the word from a request refuses it here instead.
func Reading(word string) (Counting, bool) {
	switch Counting(word) {
	case CountingReal, CountingSeeded, CountingEverybody:
		return Counting(word), true
	default:
		return CountingReal, false
	}
}

// Reached is one identity having reached one step, defined here and satisfied
// by the module that owns the stream.
type Reached func(ctx context.Context, tenantID uuid.UUID,
	names []string, since time.Time, who Counting) ([]Reach, error)

/*
ReachedAnywhere is the same question asked of the events that belong to no
school.

	IT IS A SECOND READER AND NOT A NULLABLE TENANT ON THE FIRST. `Reached` takes
	a `uuid.UUID`, and the way to say "no school" through it would be
	`uuid.Nil` — a value that means "the zero school" as readily as it means
	"every school", at a call site where getting it wrong returns an empty
	answer rather than an error. Two functions cannot be confused for each
	other, and each one's SQL says plainly which rows it is about.

	IT IS STILL SCOPED BY THE PEOPLE, not by nothing. The funnel folds what
	comes back into the same set of identities as every other step, so a
	platform event only reaches a school's chart through somebody who arrived
	at that school. The query is wide; the reading is not.
*/
type ReachedAnywhere func(ctx context.Context,
	names []string, since time.Time, who Counting) ([]Reach, error)

// Reach is the stream's answer: a step, and whichever identities were on the
// event.
type Reach struct {
	Name      string
	VisitorID *uuid.UUID
	AccountID *uuid.UUID
}

// Links is the visitor-to-account map, from the module that owns it.
type Links func(ctx context.Context) (map[uuid.UUID]uuid.UUID, error)

// The eight steps, in order, and the event each is counted from.
//
// THE ORDER IS THE PRODUCT AND NOT THE DATA. It is written here rather than
// derived from what the stream happens to contain, so that a step nobody has
// reached still appears — a funnel that hid its empty steps would hide exactly
// the drop somebody is looking for.
var steps = []Step{
	{Event: "visitor.arrived", Label: "Arrived"},
	{Event: "account.created", Label: "Created an account"},
	/* CONFIRMING IS NOT A GATE, so this step can be HIGHER than the one below
	   it and that is not a bug. Registering signs a student in and nothing waits
	   on the address being proved — somebody who never opens the mail carries
	   straight on to a track. A reader who expects a funnel to fall at every
	   step will read the rise as broken, which is why it is written here rather
	   than left to be rediscovered.

	   What the step is worth is exactly that comparison: the gap between it and
	   "Chose a track" is how many people we cannot reach among the ones who are
	   actually studying. */
	{Event: "account.confirmed", Label: "Verified the address"},
	{Event: "track.opened", Label: "Chose a track"},
	{Event: "lesson.opened", Label: "Opened the first lesson"},
	{Event: "section.completed", Label: "Finished the first section"},
	{Event: "course.completed", Label: "Finished the free course"},
	/* AND THE EIGHTH IS MEASURED NOW, which it was not since this file existed.

	   What it used to say — "nothing creates a subscription until there is a
	   payment gateway" — stopped being true when phase 3 shipped and stayed on
	   the screen anyway, which is the quietest kind of wrong: a sentence that
	   explains an absence somebody would otherwise investigate. Billing has
	   created subscriptions for weeks; what did not exist was the event, and
	   `billing.EventStarted` is it.

	   IT COUNTS STARTING AND NOT PAYING. A renewal is a different name in the
	   stream on purpose — a funnel asks how many people got this far, once, and
	   a step that counted every payment would grow every year without one more
	   person ever reaching it.

	   THE NAME IS THIS PACKAGE'S OWN AND NOT `billing.EventStarted`, which is
	   X-02: modules do not import modules. It is the arrangement every other
	   step is already under — `section.completed` is written out in `progress`
	   where it is emitted and again in `cohort.go` where it is read — because a
	   name in an append-only stream is a contract with rows written years ago
	   rather than a variable one package owns. A rename reaching only one end
	   is a query that matches nothing and a report that says nobody ever
	   subscribed, so `TestTheStreamsNamesAreWrittenAtBothEnds` holds the pair. */
	{Event: SubscribedEvent, Label: "Subscribed", Platform: true},
}

// SubscribedEvent is the name this package reads the last step from. Its other
// end is `billing.EventStarted`; see the step above for why there are two.
const SubscribedEvent = "subscription.started"

// Funnel answers how many people reached each step in one school, over the
// population `who` names.
//
// THE POPULATION IS AN ARGUMENT AND NOT A DEFAULT BURIED IN THE SQL, because
// this report is the one place the seeded students are allowed to be counted —
// a screen showing a demonstration, which says so on its face. Everything that
// ACTS stays on `CountingReal` and has no way to ask for otherwise; see
// `event.Counting`, which this mirrors.
func (s *Store) Funnel(ctx context.Context, tenantID uuid.UUID, since time.Time,
	who Counting) ([]Step, error) {
	if s.reached == nil || s.anywhere == nil || s.links == nil {
		return nil, fmt.Errorf("analysis: this store was built without the stream to read")
	}

	/* TWO LISTS, BECAUSE THEY ARE TWO QUESTIONS. Seven of these steps happened
	   inside this school and the eighth did not belong to any — see `Platform`
	   on `Step`. Asking for all eight with a tenant would drop the eighth
	   silently, which is the failure this whole file is written against: a step
	   that reports zero where it should report a number, and looks like a
	   cliff. */
	var here, anywhere []string
	for _, step := range steps {
		switch {
		case step.Event == "":
		case step.Platform:
			anywhere = append(anywhere, step.Event)
		default:
			here = append(here, step.Event)
		}
	}

	reaches, err := s.reached(ctx, tenantID, here, since, who)
	if err != nil {
		return nil, fmt.Errorf("analysis: reading who reached each step: %w", err)
	}

	beyond, err := s.anywhere(ctx, anywhere, since, who)
	if err != nil {
		return nil, fmt.Errorf("analysis: reading who subscribed: %w", err)
	}

	links, err := s.links(ctx)
	if err != nil {
		return nil, fmt.Errorf("analysis: reading which visitors belong to an account: %w", err)
	}

	// One set of people per step. A person can reach a step from several
	// browsers and on several days; the set is what makes them one.
	people := map[string]map[string]bool{}
	for _, r := range reaches {
		who := personOf(r, links)
		if who == "" {
			// An event with neither identity on it. It happened, and there is
			// nobody to count it for — dropping it is right, and counting it as
			// an anonymous person would inflate every step it appears in.
			continue
		}
		if people[r.Name] == nil {
			people[r.Name] = map[string]bool{}
		}
		people[r.Name][who] = true
	}

	/* WHO THIS SCHOOL HAS SEEN AT ALL, which is what makes the platform's
	   events into this school's number.

	   THE PLATFORM READ IS DELIBERATELY WIDE AND HAS TO BE NARROWED HERE. It
	   returns every subscription on the platform, because a subscription
	   carries no school to filter by — so folding it in with the rest, as the
	   first version of this did, would have put every subscriber on the
	   platform into every school's last step. Eight schools would each have
	   claimed the same conversions, and the number would have looked plausible
	   on all eight.

	   THE SET IS EVERY STEP AND NOT JUST THE FIRST. "Arrived" is the honest
	   denominator and it is also the event most likely to be missing — it is
	   emitted for a signed-out browser, before any account exists, and it is
	   the one a blocked script or a direct link never produces. Somebody whose
	   arrival was never recorded but who opened four lessons here is plainly
	   somebody this school has seen, and requiring the first step would drop
	   them from the last one while leaving them in the middle six. */
	seen := map[string]bool{}
	for _, step := range steps {
		if step.Event == "" || step.Platform {
			continue
		}
		for who := range people[step.Event] {
			seen[who] = true
		}
	}

	for _, r := range beyond {
		who := personOf(r, links)
		if who == "" || !seen[who] {
			continue
		}
		if people[r.Name] == nil {
			people[r.Name] = map[string]bool{}
		}
		people[r.Name][who] = true
	}

	out := make([]Step, 0, len(steps))
	for _, step := range steps {
		one := step
		if step.Event != "" {
			one.Measured = true
			one.People = len(people[step.Event])
		}
		out = append(out, one)
	}
	return out, nil
}

// personOf folds the two identities on an event into one person.
//
// AN ACCOUNT WINS OVER A VISITOR, ALWAYS. Somebody signed in on two browsers is
// one person, and the account is the identity that says so. A visitor linked to
// an account is that account even when the event carried no account id — which
// is the case that matters, because it is how an arrival on Monday and a
// subscription on Friday become the same person.
func personOf(r Reach, links map[uuid.UUID]uuid.UUID) string {
	if r.AccountID != nil {
		return "account:" + r.AccountID.String()
	}
	if r.VisitorID != nil {
		if account, linked := links[*r.VisitorID]; linked {
			return "account:" + account.String()
		}
		return "visitor:" + r.VisitorID.String()
	}
	return ""
}

File: internal/console/ui/app/screens/funnel.js
/* ==========================================================================
   The funnel — of the people who arrived at a school, how many got to each step.

   IT HAS BEEN COMPUTED EVERY NIGHT AND PRINTED INTO A LOG. `cmd/analyse` says
   so in a comment: "there is no console yet, and a report nobody can read is a
   report nobody acts on". This is the screen it was waiting for, and it is the
   first entry in `Measure` — a group the rail has had a name for and nothing in.

   # THE HARD PART IS ALREADY DONE AND IS WORTH KNOWING ABOUT

   Every number here is a count of PEOPLE, not of events and not of identities.
   The top of the funnel is browsers and the bottom is accounts, so somebody who
   arrived on Monday with no account and came back signed in on Friday has to be
   one person or the conversion rate is a ratio between two different
   populations. `analysis` folds them; this screen only draws the answer.

   # A STEP WITH NO EVENT IS DRAWN AS A SENTENCE AND NEVER AS A ZERO

   Two of the eight cannot be emitted yet — verifying an address, and
   subscribing. A bar of zero would say everybody drops out there, which would
   be this screen's most alarming finding and would be about a feature nobody
   has written. So those steps carry what is missing instead of a number, and
   they are the reason the answer has a `measured` field at all.

   # THE POPULATION IS A SWITCH, AND THE BANNER COMES FROM THE SERVER

   Seeded students are excluded from every aggregate by default (K-11), and this
   is the one screen in the platform that may be told to count them — because it
   reports and does not act. The sentence saying so is NOT written here: it
   arrives on the same answer as the numbers, so a chart of real people can never
   be drawn under a heading claiming otherwise. If the server says the banner is
   empty, there is nothing to warn about.

   # THE BARS ARE RELATIVE TO THE FIRST MEASURED STEP

   Which is the definition of a funnel: everything is a share of the people who
   arrived. Drawing each bar against the largest number instead would rescale the
   chart whenever the top step changed, and the shape is the whole point.
   ========================================================================== */

import { esc } from '../dom.js';
import { get } from '../request.js';

/* The windows on offer. Days, because the API takes days and a screen that
   offered "this quarter" would be a second calendar to keep in step. */
const WINDOWS = [
  ['0', 'Since the beginning'],
  ['30', 'Last 30 days'],
  ['90', 'Last 90 days'],
  ['365', 'Last year'],
];

const NAMES = {
  real: 'Real people',
  seeded: 'The seeded population',
  everybody: 'Everybody, real and seeded',
};

export default async function funnel(section) {
  const el = document.createElement('div');
  el.className = 'view';

  el.innerHTML =
    '<header class="view-head">' +
      '<span class="eyebrow mono">Measure</span>' +
      '<h1>The funnel</h1>' +
      '<p>Of the people who arrived at a school, how many reached each step. ' +
      'Every number is a count of people rather than of visits: somebody who ' +
      'arrived without an account and came back signed in is one person, which ' +
      'is the only reason the top and the bottom of this can be compared at ' +
      'all.</p>' +
    '</header>' +
    '<div id="body" aria-live="polite"><p class="checking">Reading…</p></div>';

  const body = el.querySelector('#body');

  let schools;
  try {
    schools = (await get('/console/api/v1/schools')).schools || [];
  } catch (e) {
    body.innerHTML = '<section class="block"><p class="none">' + esc(e.message) + '</p></section>';
    return { title: section.name, el };
  }

  if (!schools.length) {
    body.innerHTML = '<section class="block"><p class="none">There are no schools on this ' +
      'platform yet, so there is nobody to have arrived at one.</p></section>';
    return { title: section.name, el };
  }

  /* WHAT IS BEING ASKED, HELD IN ONE PLACE. Three controls change it and one
     function answers it, so a redraw cannot end up showing one school's numbers
     under another school's name. */
  const asking = { school: schools[0].id, days: '0', counting: 'real' };

  body.innerHTML =
    '<section class="block">' +
      '<div class="block-top"><h2>What to count</h2></div>' +
      '<form id="ask" class="list-bar" novalidate>' +
        '<label class="field">' +
          '<span>School</span>' +
          '<select id="school">' +
            schools.map((s) =>
              '<option value="' + esc(s.id) + '">' + esc(s.name) + '</option>').join('') +
          '</select>' +
        '</label>' +
        '<label class="field">' +
          '<span>Window</span>' +
          '<select id="days">' +
            WINDOWS.map(([value, label]) =>
              '<option value="' + value + '">' + esc(label) + '</option>').join('') +
          '</select>' +
        '</label>' +
        '<label class="field">' +
          '<span>People</span>' +
          '<select id="counting">' +
            Object.keys(NAMES).map((k) =>
              '<option value="' + k + '">' + esc(NAMES[k]) + '</option>').join('') +
          '</select>' +
        '</label>' +
      '</form>' +
    '</section>' +
    '<div id="chart"><p class="checking">Reading…</p></div>';

  const chart = body.querySelector('#chart');
  const controls = body.querySelector('#ask');

  controls.addEventListener('change', (event) => {
    const id = event.target.id;
    if (id === 'school' || id === 'days' || id === 'counting') {
      asking[id === 'school' ? 'school' : id] = event.target.value;
      draw();
    }
  });

  await draw();
  return { title: section.name, el };

  async function draw() {
    chart.innerHTML = '<p class="checking">Reading…</p>';

    /* THE REQUEST THAT WAS SENT IS REMEMBERED, so an answer that arrives after
       somebody has already changed the school is dropped rather than drawn.
       Three controls and a network make that ordinary, not exotic. */
    const mine = JSON.stringify(asking);

    let answer;
    try {
      answer = await get('/console/api/v1/schools/' + encodeURIComponent(asking.school) +
        '/funnel?days=' + encodeURIComponent(asking.days) +
        '&counting=' + encodeURIComponent(asking.counting));
    } catch (e) {
      if (mine !== JSON.stringify(asking)) return;
      chart.innerHTML = '<section class="block"><p class="none">' + esc(e.message) + '</p></section>';
      return;
    }
    if (mine !== JSON.stringify(asking)) return;

    const steps = answer.steps || [];

    /* THE TOP OF THE FUNNEL IS THE FIRST MEASURED STEP AND NOT THE BIGGEST
       NUMBER. Everything below is a share of the people who arrived, which is
       what makes this a funnel rather than a bar chart. */
    const top = steps.find((s) => s.measured);
    const arrived = top ? top.people : 0;

    chart.innerHTML =
      /* THE BANNER IS THE SERVER'S SENTENCE, drawn only when it sent one. The
         screen does not decide when the numbers are about invented students —
         the answer that carries them does, so the two cannot disagree. */
      (answer.banner
        ? '<p class="notice-strong" role="status">' + esc(answer.banner) + '</p>'
        : '') +

      '<section class="block">' +
        '<div class="block-top">' +
          '<h2>' + esc((answer.school && answer.school.name) || '') + '</h2>' +
          '<span class="block-score mono">' + esc(NAMES[answer.counting] || answer.counting) +
          '</span>' +
        '</div>' +
        (arrived === 0 && steps.every((s) => !s.measured || s.people === 0)
          ? '<p class="none">Nobody has reached any step of this, in this window, ' +
            'for these people. An empty funnel is a real answer and not a failure ' +
            'to read one.</p>'
          : '<ol class="funnel">' + steps.map((s) => row(s, arrived)).join('') + '</ol>') +
      '</section>';
  }

  function row(step, arrived) {
    /* NOT MEASURED IS NOT ZERO, and the two do not share a shape on the screen
       either — there is no bar at all, because a bar of length nothing is
       exactly the reading this is here to prevent. */
    if (!step.measured) {
      return '<li class="funnel-step funnel-step-unmeasured">' +
        '<span class="funnel-label">' + esc(step.label) + '</span>' +
        '<span class="funnel-missing">Not counted yet — ' + esc(step.why || '') + '</span>' +
      '</li>';
    }

    const share = arrived > 0 ? step.people / arrived : 0;
    return '<li class="funnel-step">' +
      '<span class="funnel-label">' + esc(step.label) + '</span>' +
      '<span class="funnel-bar"><span class="funnel-fill" ' +
        'style="width:' + (share * 100).toFixed(1) + '%"></span></span>' +
      '<span class="funnel-count mono">' + step.people +
        (arrived > 0 ? ' · ' + Math.round(share * 100) + '%' : '') +
      '</span>' +
    '</li>';
  }
}
